### DISCUSSION TITLE
How to interpret "faces optimize" command output?

### DISCUSSION THREAD

**Discussion by srett** [photoprism/photoprism#4606] [Q&A] [answered]
I've been experimenting with tweaking the parameters related to facial recognition over the past days, and have a question regarding the `faces optimize` CLI command. I've set its log level to debug, and noticed that if I do multiple consecutive runs, I get identical output every time, i.e. it's telling me about a lot of merges, and removed orphans, but on the next run I get exactly the same output. So is it not doing anything, but just doing a dry run? Or is something broken?

This is the last few lines as an example. They are identical on every run, just like the rest of the output:

```
DEBU[2024-11-03T10:06:32+01:00] faces: can merge NQARLUQS2YZXA3YPORGXE2EDQU4LU7BU with ZOJPAA2ZGTB4EHM3M7HTTNV5TGETSV2C, subject 'XXXXX', dist 0.193616 
DEBU[2024-11-03T10:06:32+01:00] faces: can merge NQARLUQS2YZXA3YPORGXE2EDQU4LU7BU with 4CZJYYLSKREUC63O57ETXE5PSI7T4BG7, subject 'XXXXX', dist 0.193616 
DEBU[2024-11-03T10:06:33+01:00] faces: removed 3 orphans for subject jsinvsqrfo99l7yx 
INFO[2024-11-03T10:06:33+01:00] merged 550 face clusters in 1m50.522155445s  
DEBU[2024-11-03T10:06:33+01:00] closed database connection  
```

This was the command I used:
```
photoprism --face-score 7 --face-cluster-size 70 --face-cluster-score 10 --face-match-dist 0.55 --log-level debug faces optimize
```

I'm not sure I need to pass all these options to the call too or whether it would read them from the config file, but I thought I just make sure to give the identical options that I launch photoprism with.

**Comment by lastzero** [comment, marked as answer]
It could be that previous changes/optimisations are later reversed each time the command is executed? Also, as you have specified custom parameters, for example ` --face-match-dist 0.55`, these may differ from the settings your instance is using, or from the settings originally used to create the clusters and match the faces. We haven't tested everything that could happen under these circumstances, so I can't confirm or deny that this is the cause.

When we have time, we will switch to a proper vector database/library, which should solve this and similar issues.

**Reply by srett** [reply]
> It could be that previous changes/optimisations are later reversed each time the command is executed?

That would be my hunch too, but it's hard to verify as this only happens on my actual instance with almost 10k face markers, not in a simple test setup.

> Also, as you have specified custom parameters, for example ` --face-match-dist 0.55`, these may differ from the settings your instance is using, or from the settings originally used to create the clusters and match the faces.

I've made sure they match what the instance is running with, but you're right in that I changed these values over time, trying to improve the automatic matching, especially with children.

> We haven't tested everything that could happen under these circumstances, so I can't confirm or deny that this is the cause.

Yeah, understandable, this seems very hard to reproduce or isolate, unfortunately.

I've also now tried to do a soft-reset of faces like so:

```sql
UPDATE markers SET face_id = '', face_dist = -1, updated_at = 0 WHERE marker_type = 'face';
TRUNCATE faces;
UPDATE markers SET subj_uid = '', marker_name = '' WHERE marker_type = 'face' AND subj_src = '';
```

This should remove all clusters and reset all markers that were assigned automatically. After that, running `faces update` again recreates clusters, but only very few. I then run the command multiple times until it does no more conflict resolution or merging. This seems to always converge after a few runs. But after tagging a few more faces, I get the same issue again, i.e. the call always finding the same conflicts over and over again. But I'm still doing more experiments here over the next couple days probably.

> When we have time, we will switch to a proper vector database/library, which should solve this and similar issues.

This sounds great. It's still performing OK-ish for me. Tagging a new face sometimes takes up to 20 seconds, but is mostly 2-3 secs.

**Reply by lastzero** [reply]
What I can say is that a distance of 0.55 is relatively large - even adding 0.01 can make a big difference here. So you end up with very large, not very specific clusters, which would then have to be broken up again to match specific faces (if they belong to different people). Because we're using a simple distance, it's a spherical geometry which, with such a large volume/radius, may not fit well with the actual data.

**Reply by srett** [reply]
I see. I might try the reverse then, increasing the cluster-dist and keeping the match-dist low, but if that doesn't work out well either, I might just revert to the defaults for the time being. :)
